### ISSUE TITLE
Stage 2 노트북 - Baseline Training

### ISSUE THREAD

**Issue by Bkankim** [Bkankim/naturallanguageprocessingcompetition-nlp-1#9]
# Task 003: Stage 2 노트북 - Baseline Training

## Description

KoBART 모델을 활용하여 대화 요약 태스크의 베이스라인 모델을 학습합니다. Hugging Face Trainer를 사용하여 효율적인 학습 파이프라인을 구축하고, ROUGE 메트릭으로 모델 성능을 평가합니다. 학습된 모델로 테스트 데이터에 대한 예측을 생성하여 첫 번째 Kaggle 제출 파일을 작성합니다.

## Acceptance Criteria

- [ ] `notebooks/stage2_baseline_training.ipynb` 노트북이 생성됨
- [ ] KoBART 모델 설정 및 로딩 완료
  - [ ] 토크나이저에 special tokens 추가
  - [ ] 모델 설정 (max_length, generation config)
- [ ] 데이터셋 준비 및 전처리
  - [ ] Hugging Face Dataset 형식으로 변환
  - [ ] 토크나이징 및 패딩 적용
  - [ ] Train/validation split (90/10)
- [ ] Trainer 설정 및 학습 실행
  - [ ] TrainingArguments 정의
  - [ ] ROUGE 메트릭 콜백 구현
  - [ ] W&B 로깅 활성화
  - [ ] 최소 3 epochs 학습
- [ ] 모델 평가 및 검증
  - [ ] Validation set에서 ROUGE 점수 계산
  - [ ] 샘플 예측 결과 확인
  - [ ] 모델 체크포인트 저장
- [ ] 테스트 데이터 예측 및 제출 파일 생성
  - [ ] `submissions/submission_baseline.csv` 생성
  - [ ] Kaggle 제출 형식 준수

#### 1. Setup & Configuration
```python
import torch
from transformers import (
    AutoTokenizer,
    AutoModelForSeq2SeqLM,
    Trainer,
    TrainingArguments,
    DataCollatorForSeq2Seq
)
from datasets import Dataset
import pandas as pd
import json
from utils import setup_wandb, compute_rouge, set_seed, auto_git_backup

set_seed(42)

# Load processed data
train_df = pd.read_csv('../data/processed/train_clean.csv')
test_df = pd.read_csv('../data/processed/test_clean.csv')

# Load special tokens
with open('../configs/special_tokens.json', 'r') as f:
    special_tokens = json.load(f)
```

#### 2. Model & Tokenizer Setup
```python
MODEL_NAME = "gogamza/kobart-base-v2"

# Initialize tokenizer
tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)

# Add special tokens
num_added = tokenizer.add_special_tokens(special_tokens)
print(f"Added {num_added} special tokens")

# Initialize model
model = AutoModelForSeq2SeqLM.from_pretrained(MODEL_NAME)

# Resize token embeddings
if num_added > 0:
    model.resize_token_embeddings(len(tokenizer))

# Model configuration
MAX_INPUT_LENGTH = 512
MAX_TARGET_LENGTH = 128
GENERATION_CONFIG = {
    "max_length": MAX_TARGET_LENGTH,
    "num_beams": 4,
    "length_penalty": 1.0,
    "early_stopping": True,
    "no_repeat_ngram_size": 3
}
```

#### 3. Dataset Preparation
```python
def preprocess_function(examples):
    """토크나이징 함수"""
    inputs = examples['dialogue_clean']
    targets = examples['summary_clean']

    model_inputs = tokenizer(
        inputs,
        max_length=MAX_INPUT_LENGTH,
        truncation=True,
        padding='max_length'
    )

    labels = tokenizer(
        targets,
        max_length=MAX_TARGET_LENGTH,
        truncation=True,
        padding='max_length'
    )

    model_inputs['labels'] = labels['input_ids']
    return model_inputs

# Create HF datasets
train_dataset = Dataset.from_pandas(train_df)
train_dataset = train_dataset.map(
    preprocess_function,
    batched=True,
    remove_columns=train_dataset.column_names
)

# Train/validation split
split_dataset = train_dataset.train_test_split(test_size=0.1, seed=42)
train_ds = split_dataset['train']
val_ds = split_dataset['test']

print(f"Train size: {len(train_ds)}, Validation size: {len(val_ds)}")
```

#### 4. Training Configuration
```python
# W&B setup
setup_wandb(
    project_name="dialogue-summarization",
    config_dict={
        "model": MODEL_NAME,
        "max_input_length": MAX_INPUT_LENGTH,
        "max_target_length": MAX_TARGET_LENGTH,
        "epochs": 3,
        "batch_size": 8,
        "learning_rate": 5e-5
    }
)

# Training arguments
training_args = TrainingArguments(
    output_dir='../checkpoints/kobart-baseline',
    evaluation_strategy='epoch',
    save_strategy='epoch',
    learning_rate=5e-5,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    num_train_epochs=3,
    weight_decay=0.01,
    save_total_limit=2,
    load_best_model_at_end=True,
    metric_for_best_model='rouge1',
    greater_is_better=True,
    logging_steps=100,
    report_to='wandb',
    fp16=torch.cuda.is_available(),
    gradient_accumulation_steps=2,
    warmup_steps=500
)

# Data collator
data_collator = DataCollatorForSeq2Seq(
    tokenizer=tokenizer,
    model=model,
    padding=True
)
```

#### 5. ROUGE Metric Implementation
```python
def compute_metrics(eval_pred):
    """ROUGE 메트릭 계산"""
    predictions, labels = eval_pred

    # Decode predictions
    decoded_preds = tokenizer.batch_decode(
        predictions,
        skip_special_tokens=True
    )

    # Replace -100 in labels
    labels = np.where(labels != -100, labels, tokenizer.pad_token_id)
    decoded_labels = tokenizer.batch_decode(
        labels,
        skip_special_tokens=True
    )

    # Compute ROUGE
    rouge_scores = compute_rouge(decoded_preds, decoded_labels)

    return {
        'rouge1': rouge_scores['rouge1'],
        'rouge2': rouge_scores['rouge2'],
        'rougeL': rouge_scores['rougeL']
    }
```

#### 6. Training Execution
```python
# Initialize Trainer
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_ds,
    eval_dataset=val_ds,
    tokenizer=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics
)

# Train
print("Starting training...")
train_result = trainer.train()

# Save final model
trainer.save_model('../checkpoints/kobart-baseline/final')
tokenizer.save_pretrained('../checkpoints/kobart-baseline/final')

# Print training results
print(f"Training completed!")
print(f"Training loss: {train_result.training_loss:.4f}")
```

#### 7. Evaluation & Sample Predictions
```python
# Evaluate on validation set
eval_results = trainer.evaluate()
print("\nValidation Results:")
for key, value in eval_results.items():
    print(f"{key}: {value:.4f}")

# Sample predictions
def generate_summary(dialogue_text, max_length=MAX_TARGET_LENGTH):
    """단일 텍스트에 대한 요약 생성"""
    inputs = tokenizer(
        dialogue_text,
        max_length=MAX_INPUT_LENGTH,
        truncation=True,
        return_tensors='pt'
    ).to(model.device)

    outputs = model.generate(
        **inputs,
        **GENERATION_CONFIG
    )

    summary = tokenizer.decode(outputs[0], skip_special_tokens=True)
    return summary

# Test on 5 random samples
sample_indices = np.random.choice(len(val_ds), 5, replace=False)
print("\nSample Predictions:")
for idx in sample_indices:
    dialogue = train_df.iloc[idx]['dialogue_clean']
    reference = train_df.iloc[idx]['summary_clean']
    prediction = generate_summary(dialogue)

    print(f"\n{'='*80}")
    print(f"Dialogue: {dialogue[:200]}...")
    print(f"Reference: {reference}")
    print(f"Prediction: {prediction}")
```

#### 8. Test Prediction & Submission
```python
# Generate predictions for test set
def generate_test_predictions(test_df, batch_size=16):
    """테스트 데이터에 대한 예측 생성"""
    predictions = []

    for i in range(0, len(test_df), batch_size):
        batch = test_df.iloc[i:i+batch_size]
        dialogues = batch['dialogue_clean'].tolist()

        # Batch tokenization
        inputs = tokenizer(
            dialogues,
            max_length=MAX_INPUT_LENGTH,
            truncation=True,
            padding=True,
            return_tensors='pt'
        ).to(model.device)

        # Generate
        outputs = model.generate(**inputs, **GENERATION_CONFIG)

        # Decode
        summaries = tokenizer.batch_decode(outputs, skip_special_tokens=True)
        predictions.extend(summaries)

        if (i // batch_size + 1) % 10 == 0:
            print(f"Processed {i+batch_size}/{len(test_df)} samples")

    return predictions

print("Generating test predictions...")
test_predictions = generate_test_predictions(test_df)

# Create submission file
submission_df = pd.DataFrame({
    'id': test_df['id'],
    'summary': test_predictions
})

submission_df.to_csv('../submissions/submission_baseline.csv', index=False)
print(f"Submission file saved: submissions/submission_baseline.csv")
print(f"Total predictions: {len(submission_df)}")
```

#### 9. Backup & Logging
```python
# Auto backup
auto_git_backup("Complete baseline training - KoBART model")

# Log final metrics to W&B
import wandb
wandb.log({
    "final_rouge1": eval_results['eval_rouge1'],
    "final_rouge2": eval_results['eval_rouge2'],
    "final_rougeL": eval_results['eval_rougeL']
})
wandb.finish()

print("\nBaseline training pipeline completed!")
```

### Expected Performance
- **Baseline ROUGE Scores** (목표):
  - ROUGE-1: 0.35-0.40
  - ROUGE-2: 0.15-0.20
  - ROUGE-L: 0.30-0.35

### Output Files
- `checkpoints/kobart-baseline/final/`: 학습된 모델 체크포인트
- `submissions/submission_baseline.csv`: Kaggle 제출 파일
- W&B run logs: 학습 과정 및 메트릭

## Dependencies

- **Depends on**:
  - Task 001 (유틸리티 함수 필요)
  - Task 002 (전처리된 데이터 및 special tokens 필요)
- **Blocks**: 후속 모델 개선 태스크들
- **Parallel execution**: No (Task 001, 002 완료 후 실행)

## Effort Estimate

- **Size**: M (Medium)
- **Estimated Hours**: 5-6 hours (학습 시간 포함)
- **Parallel**: False
- **Breakdown**:
  - 모델 및 데이터셋 설정: 60 minutes
  - Trainer 설정 및 메트릭 구현: 60 minutes
  - 모델 학습 (3 epochs): 120-180 minutes
  - 평가 및 예측 생성: 45 minutes
  - 제출 파일 생성 및 검증: 30 minutes
  - 문서화 및 백업: 15 minutes

## Definition of Done

1. 노트북이 처음부터 끝까지 에러 없이 실행됨
2. 모델이 최소 3 epochs 학습되고 수렴함
3. Validation ROUGE 점수가 기대치 내에 있음 (ROUGE-1 > 0.30)
4. 모델 체크포인트가 `checkpoints/kobart-baseline/final/`에 저장됨
5. `submissions/submission_baseline.csv`가 올바른 형식으로 생성됨:
   - 필수 컬럼: `id`, `summary`
   - 행 수가 테스트 데이터 크기와 일치
   - 결측치 없음
6. W&B에 학습 과정이 로깅됨 (loss curves, metrics)
7. 5개 샘플 예측이 노트북에 표시되고 품질 확인됨
8. Git 커밋 및 백업 완료
9. 노트북에 충분한 마크다운 설명 포함
10. 첫 Kaggle 제출 준비 완료
